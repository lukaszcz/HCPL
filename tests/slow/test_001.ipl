
// test tail recursion

let f = \x \a \&b {
   if x == 0 then
      a
   else
      f (x - 1) (a + b) b
};

let fe = \x \a \b {
   if x == 0 then
      a
   else
      fe (x - 1) (a + b) b
};

print (fe 100000 0 4);
print (fe 100000 0 2);

print (f 100000 0 4);
print (f 100000 0 2);

print (fe 10000000 0 4);
print (fe 10000000 0 2);

print (f 10000000 0 4);
print (f 10000000 0 2);
