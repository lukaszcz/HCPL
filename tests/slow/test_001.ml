
let rec f x a b =
  if x = 0 then
    a
  else
    f (x - 1) (a + (Lazy.force b)) b

let rec fe x a b =
  if x = 0 then
    a
  else
    fe (x - 1) (a + b) b;;

print_endline (string_of_int (fe 100000 0 4));
print_endline (string_of_int (fe 100000 0 2));

print_endline (string_of_int (f 100000 0 (lazy 4)));
print_endline (string_of_int (f 100000 0 (lazy 2)));

print_endline (string_of_int (fe 10000000 0 4));
print_endline (string_of_int (fe 10000000 0 2));

print_endline (string_of_int (f 10000000 0 (lazy 4)));
print_endline (string_of_int (f 10000000 0 (lazy 2)));
