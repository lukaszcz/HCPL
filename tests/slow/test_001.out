400000
200000
400000
200000
400000000
200000000
400000000
200000000
